File: datafn/core/src/aggregate.ts
/**
 * Shared aggregation computation.
 * Used by both server and client offline aggregate query execution.
 */

/**
 * Calculate an aggregation over a set of records for a given op and field.
 * Null-safe: null/undefined values are excluded from sum/min/max/avg.
 * Returns null for empty sets (SQL semantics).
 */
export function calculateAggregation(
  op: string,
  field: string,
  records: Record<string, unknown>[],
): unknown {
  if (op === "count") {
    // LOW-007 (intentional): count() always returns the total number of records,
    // regardless of the `field` parameter. This matches the client's expectation
    // ("count is not null-filtered — it counts records, not values").
    // If SQL COUNT(field) semantics (null-excluded) are needed in future,
    // add a separate `countNonNull` operation rather than changing this behaviour.
    return records.length;
  }

  const values = records
    .map((r) => r[field])
    .filter((v) => v !== null && v !== undefined);

  if (values.length === 0) return null;

  if (op === "sum") {
    return values.reduce((a, b) => (Number(a) || 0) + (Number(b) || 0), 0);
  }
  if (op === "min") {
    let min = values[0];
    for (const v of values) {
      if (v < min) min = v;
    }
    return min;
  }
  if (op === "max") {
    let max = values[0];
    for (const v of values) {
      if (v > max) max = v;
    }
    return max;
  }
  if (op === "avg") {
    const sum = values.reduce((a, b) => (Number(a) || 0) + (Number(b) || 0), 0);
    return (sum as number) / values.length;
  }
  return null;
}

File: datafn/python/datafn/utils/aggregate.py
from typing import Any, Dict, List, Optional


def calculate_aggregation(op: str, field: str, records: List[Dict[str, Any]]) -> Any:
    if op == "count":
        return len(records)

    values = [r.get(field) for r in records if r.get(field) is not None]

    if not values:
        return None

    if op == "sum":
        return sum(values)
    elif op == "avg":
        return sum(values) / len(values)
    elif op == "min":
        return min(values)
    elif op == "max":
        return max(values)
    else:
        raise ValueError(f"Unknown aggregation operation: {op}")
